src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,187 @@
+            }
+
+            protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+                => await MoveFileToMatchNamespaceAsync(cancellationToken).ConfigureAwait(false);
+
+            private async Task<ImmutableArray<CodeActionOperation>> MoveFileToMatchNamespaceAsync(CancellationToken cancellationToken)
+            {
+                var id = _state.OriginalDocumentId;
+                var solution = _state.Solution;
+                var document = solution.GetDocument(id);
+                var newDocumentId = DocumentId.CreateNewId(document.Project.Id, document.Name);
+
+                solution = solution.RemoveDocument(id);
+
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                solution = solution.AddDocument(newDocumentId, document.Name, text, folders: _newfolders);
+
+                return ImmutableArray.Create<CodeActionOperation>(
+                    new ApplyChangesOperation(solution),
+                    new OpenDocumentOperation(newDocumentId, activateIfAlreadyOpen: true));
+            }
+            
+            public static ImmutableArray<MoveFileCodeAction> Create(State state)
+            {
+                // Since all documents have identical folder structure, we can do the computation on any of them.
+                var document = state.Solution.GetDocument(state.DocumentIds.First());
+                var parts = state.RelativeDeclaredNamespace.Split(new[] { '.' }).ToImmutableArray();
+
+                // Invalid char can only appear in namespace name when there's error,
+                // which we have checked before creating any code actions.
+                Debug.Assert(parts.Any(s => s.IndexOfAny(Path.GetInvalidPathChars()) < 0));
+
+                var projectRootFolder = FolderInfo.CreateFolderHierarchyForProject(document.Project);
+                var candidateFolders = FindCandidateFolders(projectRootFolder, parts, ImmutableArray<string>.Empty);
+                return candidateFolders.SelectAsArray(folders => new MoveFileCodeAction(state, folders));
+            }
+
+            /// <summary>
+            /// We try to provide additional "move file" options if we can find existing folders that matches target namespace.
+            /// For example, if the target namespace is 'DefaultNamesapce.A.B.C', and there's a folder 'ProjectRoot\A.B\' already 
+            /// exists, then will provide two actions, "move file to ProjectRoot\A.B\C\" and "move file to ProjectRoot\A\B\C\".
+            /// </summary>
+            private static ImmutableArray<ImmutableArray<string>> FindCandidateFolders(
+                FolderInfo currentFolderInfo, 
+                ImmutableArray<string> parts, 
+                ImmutableArray<string> currentFolder)
+            {
+                if (parts.Length == 0 || parts.Length == 1 && parts[0].Length == 0)
```

Comment:
i find this condition super confusing.  why would you have one side vs the other?  also, the lack of parens makes precedence wonky.

Reply:
same concerns about predence applies.